File: apps/cs-master/lib/bris-parser-py/src/bris_parser/integrated.py
"""
BRIS 통합 페이지 (complain_reference_list.asp) HTML → 구조화된 dict 리스트

JS 파서(`lib/bris-parser/src/integrated.js`)와 동등성 보장:
- projectClosed 는 빨간색 span 만 추출 (없으면 '미적용')
- echo_id/customer_id 는 '없음' → '' (None 아님)
- 사내/외부 강사 합산을 '강사' 키에도 저장 (하위호환)
- 구버전 "강사:" 단일 라벨도 지원
"""

import re
from bs4 import BeautifulSoup

from .label_values import extract_label_values
from .utils import normalize_team_name, normalize_date


class BrisParser:
    """BRIS 컴플레인참조 HTML → 한글 키 dict 리스트"""

    # 내보내기 시 사용할 컬럼 순서
    COLUMNS = [
        'business_id', 'project_id', 'echo_id', 'place_id', 'customer_id',
        '과정명', '프로그램명', '과정_총매출', '시작일', '종료일',
        '대면_비대면', '수주코드', '수주_프로젝트명', '수주_프로젝트_등록일',
        '수주일', '수주_프로젝트_마감일',
        '에코_상태', '에코_제외_사유',
        '수주_담당자', '수주팀', '수행_담당자', '수행팀',
        '사내강사', '외부강사', '강사',
        '회사명', '사업자번호', '사업장명',
        '고객_담당자', '고객_부서', '고객_이메일', '고객_전화', '고객_휴대폰',
    ]

    @staticmethod
    def parse(html_content: str) -> list:
        soup = BeautifulSoup(html_content, 'html.parser')
        table = soup.find('table', class_='ntbl_list_c')
        if not table:
            return []

        rows = table.find_all('tr')
        records = []
        i = 0

        while i < len(rows):
            row = rows[i]

            # 헤더/info-row 건너뛰기
            row_classes = row.get('class', [])
            if 'tfirst' in row_classes or 'info-row' in row_classes:
                i += 1
                continue

            tds = row.find_all('td', recursive=False)
            if len(tds) < 5 or any(td.get('colspan') for td in tds):
                i += 1
                continue

            try:
                record = BrisParser._parse_main_row(tds)
            except Exception:
                i += 1
                continue

            # 후속 info-row 수집 (최대 10개 — 사내/외부 강사가 별개 행일 수 있음)
            info_rows = []
            j = i + 1
            while j < len(rows) and j <= i + 10:
                nr = rows[j]
                nr_tds = nr.find_all('td', recursive=False)
                if nr_tds and nr_tds[0].get('colspan'):
                    info_rows.append(nr)
                    j += 1
                else:
                    break

            BrisParser._parse_info_rows(record, info_rows)

            # 하위호환: 사내+외부 강사 합산을 '강사'(단일 키)에도 보관
            combined = ', '.join(
                v for v in [record.get('사내강사', ''), record.get('외부강사', '')]
                if v
            )
            if combined and not record.get('강사'):
                record['강사'] = combined

            records.append(record)
            i = j

        return records

    @staticmethod
    def _parse_main_row(tds) -> dict:
        rec = {}
        # ID 추출 — '없음' 은 '' 로 (JS 동등)
        for span in tds[0].find_all('span'):
            t = span.get_text(strip=True)
            for key in ('business_id', 'project_id', 'echo_id', 'place_id', 'customer_id'):
                if key in t:
                    val = t.split(':')[-1].strip()
                    rec[key] = '' if val == '없음' else val

        rec['과정명'] = tds[1].get_text(strip=True)
        rec['프로그램명'] = tds[2].get_text(strip=True)

        revenue = tds[3].get_text(strip=True).replace(',', '')
        try:
            rec['과정_총매출'] = int(revenue) if revenue else 0
        except ValueError:
            rec['과정_총매출'] = 0

        rec['시작일'] = normalize_date(tds[4].get_text(strip=True))
        rec['종료일'] = normalize_date(tds[5].get_text(strip=True))
        rec['대면_비대면'] = tds[6].get_text(strip=True)
        rec['수주코드'] = tds[7].get_text(strip=True)
        rec['수주_프로젝트명'] = tds[8].get_text(strip=True)
        rec['수주_프로젝트_등록일'] = tds[9].get_text(strip=True)
        rec['수주일'] = tds[10].get_text(strip=True)

        # 수주_프로젝트_마감일: 빨간색 span 안의 값만 마감일로 간주 (JS 동등)
        # 없으면 '미적용'
        if len(tds) > 11:
            close_td = tds[11]
            red = close_td.find('span', style=lambda s: s and 'red' in str(s))
            if red and red.get_text(strip=True):
                rec['수주_프로젝트_마감일'] = red.get_text(strip=True)
            else:
                rec['수주_프로젝트_마감일'] = '미적용'
        else:
            rec['수주_프로젝트_마감일'] = '미적용'

        return rec

    # BRIS info-row 라벨 → record dict 키 매핑.
    # 새 라벨이 HTML 에 등장해도 _extra_labels 에 수집되므로 관찰 후 여기 추가.
    _LABEL_MAP = {
        '에코 상태':      '에코_상태',
        '에코 제외 사유': '에코_제외_사유',
        '수주':           '수주_담당자',
        '수주팀':         '수주팀',
        '수행':           '수행_담당자',
        '수행팀':         '수행팀',
        '사내강사':       '사내강사',
        '외부강사':       '외부강사',
        '회사명':         '회사명',
        '사업자번호':     '사업자번호',
        '사업장명':       '사업장명',
        '담당자':         '고객_담당자',
        '부서':           '고객_부서',
        '이메일':         '고객_이메일',
        '전화':           '고객_전화',
        '휴대':           '고객_휴대폰',
    }

    @staticmethod
    def _parse_info_rows(record, info_rows):
        """각 info-row 에서 <b>라벨:</b>값 을 일괄 추출해 record 에 매핑.

        특수 처리:
          - '에코 상태' 는 빨간색 span 이 감싼 "에코 제외" 텍스트가 우선.
          - '수주팀'/'수행팀' 은 소괄호 제거 후 normalize.
          - 구버전 단일 '강사:' 라벨은 정규식 fallback.
          - 매핑에 없는 신규 라벨은 record['_extra_labels'] 에 관찰용 보관.
        """
        extra: dict[str, str] = {}
        for row in info_rows:
            lv = extract_label_values(row)
            text = row.get_text()

            # 에코 상태: 빨간 span 우선 override
            if '에코 상태' in lv:
                red = row.find('span', style=lambda s: s and 'red' in str(s))
                if red and '에코 제외' in red.get_text():
                    lv['에코 상태'] = '에코 제외'

            for label, value in lv.items():
                field = BrisParser._LABEL_MAP.get(label)
                if field is None:
                    extra[label] = value
                    continue
                if field == '수주팀':
                    value = value.strip().strip('()')
                elif field == '수행팀':
                    value = normalize_team_name(value.strip().strip('()'))
                record[field] = value

            # 구버전 단일 "강사:" 라벨 (사내/외부강사 둘 다 없을 때만)
            if ('강사 :' in text
                    and '사내강사' not in text
                    and '외부강사' not in text
                    and not record.get('강사')):
                m = re.search(r'강사\s*:\s*(.+?)$', text, re.M)
                if m:
                    record['강사'] = m.group(1).strip()

        if extra:
            record['_extra_labels'] = extra

File: apps/cs-master/lib/bris-parser/src/integrated.js
/**
 * BRIS 통합 페이지 (complain_reference_list.asp) HTML → 과정 레코드 배열 추출
 */

import { getDOMParser } from './dom.js';
import { normalizeDate, normalizeTeamName } from './utils.js';
import { extractLabelValues } from './label-values.js';

export function extractIntegratedData(html) {
  const DOMParser = getDOMParser();
  const parser = new DOMParser();
  const doc = parser.parseFromString(html, 'text/html');

  const table = doc.querySelector('table.ntbl_list_c');
  if (!table) return [];

  const allRows = table.querySelectorAll('tr');
  const records = [];
  let i = 0;

  // 첫 번째 행(tfirst 헤더) 건너뛰기
  while (i < allRows.length && allRows[i].classList.contains('tfirst')) i++;

  function isInfoRow(tr) {
    if (tr.classList.contains('info-row')) return true;
    const firstTd = tr.querySelector('td');
    return firstTd && firstTd.classList.contains('info-row');
  }

  while (i < allRows.length) {
    const row = allRows[i];
    if (isInfoRow(row)) { i++; continue; }

    const tds = row.querySelectorAll('td');
    if (tds.length < 12) { i++; continue; }

    // 첫 번째 td에서 ID들 추출
    const idCell = tds[0].textContent;
    const bizMatch = idCell.match(/business_id\s*:\s*(\d+)/);
    const projMatch = idCell.match(/project_id\s*:\s*(\d+)/);
    const echoMatch = idCell.match(/echo_id\s*:\s*(\d+)/);
    const placeMatch = idCell.match(/place_id\s*:\s*(\d+)/);
    const custMatch = idCell.match(/customer_id\s*:\s*(\d+)/);

    const businessId = bizMatch ? bizMatch[1] : '';
    const projectId = projMatch ? projMatch[1] : '';
    const echoId = echoMatch ? echoMatch[1] : '';
    const placeId = placeMatch ? placeMatch[1] : '';
    const customerId = custMatch ? custMatch[1] : '';

    const courseName = tds[1].textContent.trim();
    const programName = tds[2].textContent.trim();
    const totalRevenue = tds[3].textContent.trim();
    const startDate = normalizeDate(tds[4].textContent.trim());
    const endDate = normalizeDate(tds[5].textContent.trim());
    const eduDelivery = tds[6].textContent.trim().replace(/\s+/g, '');
    const orderCode = tds[7].textContent.trim();
    const projectName = tds[8].textContent.trim();
    const registrationDate = tds[9].textContent.trim();
    const orderDate = tds[10].textContent.trim();

    // 마감일: 빨간색 span이 있으면 마감됨
    const closeDateTd = tds[11];
    const redSpan = closeDateTd.querySelector('span[style*="color:red"], span[style*="color: red"]');
    const projectClosed = (redSpan && redSpan.textContent.trim()) ? redSpan.textContent.trim() : '미적용';

    const record = {
      businessId, projectId, echoId, placeId, customerId,
      courseName, programName, totalRevenue, startDate, endDate,
      eduDelivery, orderCode, projectName, registrationDate, orderDate,
      projectClosed,
      echoStatus: '', echoExcludeReason: '',
      am: '', amTeam: '', performer: '', performerTeam: '',
      instructor: '', internalInstructors: '', externalInstructors: '',
      company: '', businessNumber: '', placeName: '',
      dmName: '', dmDept: '', dmEmail: '', dmPhone: '', dmMobile: ''
    };

    // info-row 들을 라벨 기반으로 파싱 (사내/외부 강사가 별도 행으로 존재할 수 있음)
    i++;
    while (i < allRows.length) {
      const infoRow = allRows[i];
      if (!isInfoRow(infoRow)) break;

      const infoText = infoRow.textContent;
      const lv = extractLabelValues(infoRow);

      if (infoText.includes('에코 상태')) {
        if (lv['에코 상태']) record.echoStatus = lv['에코 상태'];
        else {
          const statusMatch = infoText.match(/에코 상태\s*:\s*(.+?)$/m);
          if (statusMatch) record.echoStatus = statusMatch[1].trim();
        }
        const echoRedSpan = infoRow.querySelector('span[style*="color: red"], span[style*="color:red"]');
        if (echoRedSpan && echoRedSpan.textContent.includes('에코 제외')) {
          record.echoStatus = '에코 제외';
        }
        // 제외 사유 — 같은 info-row 내 별도 <b>에코 제외 사유 : </b> 라벨
        if (lv['에코 제외 사유']) {
          record.echoExcludeReason = lv['에코 제외 사유'];
        } else if (infoText.includes('에코 제외 사유')) {
          const reasonMatch = infoText.match(/에코 제외 사유\s*:\s*(.+?)(?:\n|$)/);
          if (reasonMatch) record.echoExcludeReason = reasonMatch[1].trim();
        }
      } else if (infoText.includes('수주 :') && infoText.includes('수주팀')) {
        // 1차: extractLabelValues 결과 우선 — 키가 있으면 빈 문자열이라도 존중
        if ('수주' in lv) record.am = lv['수주'];
        if ('수주팀' in lv) record.amTeam = lv['수주팀'].replace(/^\(|\)$/g, '');
        if ('수행' in lv) record.performer = lv['수행'];
        if ('수행팀' in lv) {
          const perfTeam = lv['수행팀'].replace(/^\(|\)$/g, '');
          if (perfTeam) record.performerTeam = normalizeTeamName(perfTeam);
        }
        // 2차: lv 에 키 자체가 없을 때만 정규식 fallback
        // (?!팀) 음성 전방탐색 — '수주팀' / '수행팀' 오매치 방지
        if (!('수주' in lv)) {
          const amMatch = infoText.match(/수주(?!팀)\s*:\s*(\S+)/);
          if (amMatch) record.am = amMatch[1].trim();
        }
        if (!('수주팀' in lv)) {
          const amTeamMatch = infoText.match(/수주팀\s*:\s*\(([^)]+)\)/);
          if (amTeamMatch) record.amTeam = amTeamMatch[1].trim();
        }
        if (!('수행' in lv)) {
          const perfMatch = infoText.match(/수행(?!팀)\s*:\s*(\S+)/);
          if (perfMatch) record.performer = perfMatch[1].trim();
        }
        if (!('수행팀' in lv)) {
          const perfTeamMatch = infoText.match(/수행팀\s*:\s*\(([^)]+)\)/);
          if (perfTeamMatch) record.performerTeam = normalizeTeamName(perfTeamMatch[1].trim());
        }
      } else if (infoText.includes('사내강사') && !infoText.includes('외부강사')) {
        record.internalInstructors = lv['사내강사'] || '';
        if (!record.internalInstructors) {
          const m = infoText.match(/사내강사\s*:\s*(.+?)$/m);
          if (m) record.internalInstructors = m[1].trim();
        }
      } else if (infoText.includes('외부강사') && !infoText.includes('사내강사')) {
        record.externalInstructors = lv['외부강사'] || '';
        if (!record.externalInstructors) {
          const m = infoText.match(/외부강사\s*:\s*(.+?)$/m);
          if (m) record.externalInstructors = m[1].trim();
        }
      } else if (infoText.includes('사내강사') && infoText.includes('외부강사')) {
        record.internalInstructors = lv['사내강사'] || '';
        record.externalInstructors = lv['외부강사'] || '';
        if (!record.internalInstructors) {
          const m = infoText.match(/사내강사\s*:\s*(.+?)(?=\s*외부강사|$)/);
          if (m) record.internalInstructors = m[1].trim();
        }
        if (!record.externalInstructors) {
          const m = infoText.match(/외부강사\s*:\s*(.+?)$/m);
          if (m) record.externalInstructors = m[1].trim();
        }
      } else if (infoText.includes('강사') && !infoText.includes('사내강사') && !infoText.includes('외부강사')) {
        // 구버전 호환: 단일 "강사" 라벨만 있는 경우
        const instrMatch = infoText.match(/강사\s*:\s*(.+?)$/m);
        if (instrMatch) record.instructor = instrMatch[1].trim();
      } else if (infoText.includes('회사명') && infoText.includes('사업자번호')) {
        record.company = lv['회사명'] || '';
        record.businessNumber = lv['사업자번호'] || '';
        record.placeName = lv['사업장명'] || '';
        if (!record.company) {
          const compMatch = infoText.match(/회사명\s*:\s*(.+?)(?=\s*사업자번호|$)/);
          if (compMatch) record.company = compMatch[1].trim();
        }
        if (!record.businessNumber) {
          const bizNumMatch = infoText.match(/사업자번호\s*:\s*(\S*)/);
          if (bizNumMatch) record.businessNumber = bizNumMatch[1].trim();
        }
        if (!record.placeName) {
          const placeMatch = infoText.match(/사업장명\s*:\s*(.+?)$/m);
          if (placeMatch) record.placeName = placeMatch[1].trim();
        }
      } else if (infoText.includes('담당자') && infoText.includes('이메일')) {
        record.dmName = lv['담당자'] || '';
        record.dmDept = lv['부서'] || '';
        record.dmEmail = lv['이메일'] || '';
        record.dmPhone = lv['전화'] || '';
        record.dmMobile = lv['휴대'] || '';
        if (!record.dmName) {
          const nameMatch = infoText.match(/담당자\s*:\s*(.+?)(?=\s*(?:부서|이메일|전화|휴대|$))/);
          if (nameMatch && nameMatch[1].trim()) record.dmName = nameMatch[1].trim();
        }
        if (!record.dmDept) {
          const deptMatch = infoText.match(/부서\s*:\s*(.+?)(?=\s*(?:이메일|전화|휴대|$))/);
          if (deptMatch && deptMatch[1].trim()) record.dmDept = deptMatch[1].trim();
        }
        if (!record.dmEmail) {
          const emailMatch = infoText.match(/이메일\s*:\s*(.+?)(?=\s*(?:전화|휴대|$))/);
          if (emailMatch && emailMatch[1].trim()) record.dmEmail = emailMatch[1].trim();
        }
        if (!record.dmPhone) {
          const phoneMatch = infoText.match(/전화\s*:\s*([\d\-\s]+)/);
          if (phoneMatch && phoneMatch[1].trim()) record.dmPhone = phoneMatch[1].trim();
        }
        if (!record.dmMobile) {
          const mobileMatch = infoText.match(/휴대\s*:\s*([\d\-\s]+)/);
          if (mobileMatch && mobileMatch[1].trim()) record.dmMobile = mobileMatch[1].trim();
        }
      }

      i++;
    }

    // 하위호환: instructor (단일 필드)에 사내+외부 합산
    const combinedInstr = [record.internalInstructors, record.externalInstructors]
      .filter(Boolean).join(', ');
    if (combinedInstr) record.instructor = combinedInstr;

    records.push(record);
  }

  return records;
}
